gossip/src/cluster_info.rs

```diff
@@ -1767,82 +1767,63 @@ impl ClusterInfo {
-        let shuffle =
-            WeightedShuffle::<u64>::new("handle-pull-requests", &scores).shuffle(&mut rng);
-        let mut total_bytes = 0;
-        let mut sent_pull_responses = 0;
-        let mut sent_crds_values = 0;
-        for chunk in shuffle.map(|i: usize| &chunks[i]) {
-            let Chunk { addr, values, .. } = chunk;
-            let chunk_values: Vec<CrdsValue> = values.iter().map(|&v| v.clone()).collect();
-            let response = Protocol::PullResponse(self_id, chunk_values);
-            match Packet::from_data(Some(addr), response) {
-                Err(err) => {
-                    error!("failed to write pull-response packet: {:?}", err);
-                }
-                Ok(packet) => {
-                    let packet_size = packet.meta().size;
-                    if self.outbound_budget.take(packet_size) {
-                        total_bytes += packet_size;
-                        packet_batch.push(packet);
-                        sent_pull_responses += 1;
-                        sent_crds_values += values.len();
+        let (scores, mut pull_responses): (Vec<_>, Vec<_>) = addrs
+            .iter()
+            .zip(pull_responses)
+            .flat_map(|(addr, values)| {
+                num_crds_values += values.len();
+                split_gossip_messages(PULL_RESPONSE_MAX_PAYLOAD_SIZE, values).map(move |values| {
+                    let score = values.iter().map(get_score).max().unwrap_or_default();
+                    (score, (addr, values))
+                })
+            })
+            .collect();
+        let (total_bytes, sent_crds_values, sent_pull_responses) =
+            WeightedShuffle::new("handle-pull-requests", scores)
+                .shuffle(&mut rng)
+                .filter_map(|k| {
+                    let (addr, values) = &mut pull_responses[k];
+                    let num_values = values.len();
+                    let response = Protocol::PullResponse(self_id, std::mem::take(values));
+                    let packet = Packet::from_data(Some(addr), response)
+                        .inspect_err(|err| error!("failed to write pull-response packet: {err:?}"))
+                        .ok()?;
+                    Some((packet, num_values))
+                })
+                .take_while(|(packet, _)| {
+                    if self.outbound_budget.take(packet.meta().size) {
+                        true
                     } else {
                         self.stats.gossip_pull_request_no_budget.add_relaxed(1);
-                        break;
+                        false
                     }
-                }
-            }
-        }
+                })
+                .map(|(packet, num_values)| {
+                    let num_bytes = packet.meta().size;
+                    packet_batch.push(packet);
+                    (num_bytes, num_values)
+                })
+                .fold((0, 0, 0), |a, b| (a.0 + b.0, a.1 + b.1, a.2 + 1));
```

Comment:
`.fold()` always throws my brain for a loop. so `a` stores the accumulation for the gossip stats `total_bytes`, `sent_crds_values`, and `sent_pull_responses`, and `b` stores `num_bytes` and `num_values` from the iterator.

I would suggest using more descriptive variable names here. For example:

```suggestion
                .fold((0, 0, 0), |gossip_stats_acc, (num_bytes, num_values)| (gossip_stats_acc.0 + num_bytes, gossip_stats_acc.1 + num_values, gossip_stats_acc.2 + 1));
```
^ or you can just make `gossip_stats_acc` -> `stats_acc`

Reply:
how strongly do you feel about this?

imo the suggested code is verbose and actually less readable.
basically I am just summing up tuples here, so I find `a + b` syntax more readable.

anyways, these are just doing some metrics update, so are not as important as the protocol logic.
and I would rather not to rerun the CI again (which has become pretty slow due to flaky tests).

so i would suggest we keep merge current form, and if you want to update in a future change, that is fine.

Reply:
ya i would feel strongly about this if it was protocol logic. but ya for metric accumulation not the end of the world. i just personally don't like variable names that don't describe the underlying thing they represent. it makes reading the code harder.